File: src/skylens_model/utils/geo.py
"""GPS <-> local ENU (East/North/Up) coordinate conversion.

Pure-Python mirror of `src/skylens_core/geo.ts`. Detections and splat chunks
produced by this package must line up with the client's rendered scene, and
the client operates in a local ENU meter frame anchored at a reference GPS
point (1 scene unit = 1 meter).

IMPORTANT: this module must stay numerically in sync with geo.ts. Same
equirectangular small-area approximation, same Earth radius constant, and
the same scene axis convention: x = East, y = Up, z = -North. If you change
the formulas here, change geo.ts too (and vice versa).
"""

from __future__ import annotations

import math
from dataclasses import dataclass

_R = 6378137.0  # Earth radius (m), matches geo.ts
_DEG = math.pi / 180.0


@dataclass(slots=True)
class Gps:
    """Matches the `Gps` shape in src/skylens_core/geo.ts."""

    lat: float
    lon: float
    alt: float  # meters, ellipsoidal/relative — used as-is for Up


# Reference point that defines the local ENU origin. Alias for clarity,
# matching `GeoAnchor` in geo.ts.
GeoAnchor = Gps


@dataclass(slots=True)
class Enu:
    """Local East/North/Up offset in meters from a `GeoAnchor`."""

    e: float
    n: float
    u: float


def gps_to_enu(gps: Gps, anchor: GeoAnchor) -> Enu:
    """GPS -> local ENU meters (equirectangular small-area approximation)."""
    d_lat = (gps.lat - anchor.lat) * _DEG
    d_lon = (gps.lon - anchor.lon) * _DEG
    return Enu(
        e=d_lon * _R * math.cos(anchor.lat * _DEG),
        n=d_lat * _R,
        u=gps.alt - anchor.alt,
    )


def enu_to_gps(enu: Enu, anchor: GeoAnchor) -> Gps:
    """Local ENU meters -> GPS (inverse of `gps_to_enu`)."""
    return Gps(
        lat=anchor.lat + (enu.n / _R) / _DEG,
        lon=anchor.lon + (enu.e / (_R * math.cos(anchor.lat * _DEG))) / _DEG,
        alt=anchor.alt + enu.u,
    )


# Scene convention (matches the client's Three.js right-handed frame):
# x = East, y = Up, z = -North.
SceneVec3 = tuple[float, float, float]


def enu_to_scene(enu: Enu) -> SceneVec3:
    return (enu.e, enu.u, -enu.n)


def scene_to_enu(v: SceneVec3) -> Enu:
    return Enu(e=v[0], u=v[1], n=-v[2])


def gps_to_scene(gps: Gps, anchor: GeoAnchor) -> SceneVec3:
    return enu_to_scene(gps_to_enu(gps, anchor))


def scene_to_gps(v: SceneVec3, anchor: GeoAnchor) -> Gps:
    return enu_to_gps(scene_to_enu(v), anchor)

File: src/shared/geo.ts
// Real-world coordinates. The app operates in a local ENU (East/North/Up) meter
// frame anchored at a reference GPS point, and the Three.js scene maps 1 unit = 1
// meter. This keeps drone routes, telemetry, detections, and splat alignment all
// expressible in real GPS while the renderer stays in a simple metric frame.
//
// Kept Three-free (skylens_core is pure): scene coords are returned as Vec3 tuples.

import type { Vec3 } from './types.ts';

export interface Gps {
  lat: number;
  lon: number;
  /** Altitude in meters (ellipsoidal/relative — used as-is for Up). */
  alt: number;
}

/** Reference point that defines the local ENU origin. */
export type GeoAnchor = Gps;

export interface Enu {
  e: number;
  n: number;
  u: number;
}

const R = 6378137; // Earth radius (m)
const DEG = Math.PI / 180;

/** GPS -> local ENU meters (equirectangular small-area approximation). */
export function gpsToEnu(gps: Gps, anchor: GeoAnchor): Enu {
  const dLat = (gps.lat - anchor.lat) * DEG;
  const dLon = (gps.lon - anchor.lon) * DEG;
  return {
    e: dLon * R * Math.cos(anchor.lat * DEG),
    n: dLat * R,
    u: gps.alt - anchor.alt,
  };
}

/** Local ENU meters -> GPS (inverse of gpsToEnu). */
export function enuToGps(enu: Enu, anchor: GeoAnchor): Gps {
  return {
    lat: anchor.lat + (enu.n / R) / DEG,
    lon: anchor.lon + (enu.e / (R * Math.cos(anchor.lat * DEG))) / DEG,
    alt: anchor.alt + enu.u,
  };
}

// Scene convention (Three.js, right-handed): x = East, y = Up, z = -North.
export function enuToScene(enu: Enu): Vec3 {
  return [enu.e, enu.u, -enu.n];
}

export function sceneToEnu(v: Vec3): Enu {
  return { e: v[0], u: v[1], n: -v[2] };
}

export function gpsToScene(gps: Gps, anchor: GeoAnchor): Vec3 {
  return enuToScene(gpsToEnu(gps, anchor));
}

export function sceneToGps(v: Vec3, anchor: GeoAnchor): Gps {
  return enuToGps(sceneToEnu(v), anchor);
}
